### ISSUE TITLE
Want cargo publish --idempotent

### ISSUE THREAD

**Issue by ijackson** [rust-lang/cargo#13397]
### Problem

Sometimes a cargo publish run of multiple packages will fail.  It is then necessary to restart it.

It would be nice if we could run the whole publication from the top, and simply skip crates that were already published.  There are probably other uses.

### Proposed Solution

`cargo publish --idempotent`

If the current version of the crate has already been published, skips the crate and exits with a success status.

(Ideally it would download the crate and check it's identical to what is proposed to publish now, but I think that crates.io does unpredictable-to-cargo mangling of the uploaded data which makes such a check infeasible.)

**Comment by ijackson** [comment]
> Could you clarify the use case? Did you mean one `cargo publish` invocation to publish multiple packages? `cargo publish` doesn't support that at the moment. Workspace publish is a feature tracked [here](https://github.com/rust-lang/cargo/issues/1169).

Yes.  Because cargo can't publish a workspace in one go, we all have add-on tools of varying sophistication to publish our multi-crate workspaces.  The add-on tools iterate over the crates in a workspace, calling cargo publish for each one.

I think #1169 looks quite hard.  It has been open for a very long time and although there's work on it, a good solution there will be a little while.

One of the ways that these add-on tools are suboptimal is that after a failure of any kind, the workspace can be left "half-published".  Recovering from such a situation involves either complicated predictive logic in the tool, or manual tweaking of intermediate publication lists, etc.

So in the meantime, while #1169 is worked on, our add-on tools would be simpler and more reliable of cargo offered an idempotent publish operation.

And, I think any reasonable solution to #1169 needs to cope with failures halfway through, and for that situation to be recoverable.  That would mean that to implement #1169, cargo needs an idempotent publish operation for its own use anyway.  Ie, the underlying functionality in my request here is a necessary *part* of #1169.

But it is also separately useful, so it's worth exposing as an operation to external callers.  The notion is fairly well-defined and, I think, not that hard to implement.

> > If the current version of the crate has already been published, skips the crate and exits with a success status.
> 
> There is a similar request #3662. I wonder if we should keep that and close this one.

#3662 requests early *failure*.  I'm requesting treating "this one is already published" as a *success*.  Similar underlying logic might be involved with both, but the overall semantics are rather different.

*Failure* on republish is precisely the opposite of what is needed for reliable publication of a whole workspace (whether as part of a first-class cargo feature for this, or as part of an add-on tool that invokes cargo).  Since, after a failure, one ought to stop.  Whereas when publication of a whole workspace, "you already did this one" should mean "carry on".

**Comment by epage** [comment]
> So in the meantime, while https://github.com/rust-lang/cargo/issues/1169 is worked on, our add-on tools would be simpler and more reliable of cargo offered an idempotent publish operation.

As an author of such an add-on tool (`cargo-release`), I don't see how having this would make the tool much simpler or more reliable.

But I also feel automatic recovery is a tricky enough subject that I give the underlying steps to the user so they can decide how they want to recover, which includes a `cargo release publish --unpublished` command.

> And, I think any reasonable solution to https://github.com/rust-lang/cargo/issues/1169 needs to cope with failures halfway through, and for that situation to be recoverable. That would mean that to implement https://github.com/rust-lang/cargo/issues/1169, cargo needs an idempotent publish operation for its own use anyway. Ie, the underlying functionality in my request here is a necessary part of https://github.com/rust-lang/cargo/issues/1169.

In terms of core functionality, I see workspace publishing as the next step and we can better survey where things are at and expand from there.  Having workspace-publishing in will help us better evaluate how this mode should work rather than designing this most first and then hoping it remains coherent.

**Comment by epage** [comment]
Had some more conversations with people about this and a problem that came up is that if a version is already in the index, its unclear if its already published or the user forgot to bump the version. At least in this proposal, it requires opt-in.

**Comment by ijackson** [comment]
> Had some more conversations with people about this and a problem that came up is that if a version is already in the index, its unclear if its already published or the user forgot to bump the version. At least in this proposal, it requires opt-in.

Indeed so.  Absolutely, it has to be an opt-in for this reason.

**Comment by epage** [comment]
In discussing #15006, a thought that might be relevant here:

> All three options involve magic in a way and a concern I have with that is the defaults for cargo publish: the default is unreversable and there is no confirmation prompt. If we could go back and change it, I would make --dry-run the default and have a confirmation prompt on --execute (like cargo release). In this state, erroring out on unexpected input is much safer than a "guess my intent flag" as users have to go out of there way to be able to review what will happen.

**Comment by weihanglo** [comment]
With `-Zpackage-workspace` this becomes more relevant. At least if it failed in the middle suddenly workspace publish becomes useless. To publish again for a large workspace you'll need to exclude `publish=false` and already published packages.

**Comment by MOZGIII** [comment]
Two cents here: I'd suggest `--skip-already-published` instead of `--idempotent` since that's a bit more specific in this context.

This is, however, *not the same*. If we *actually* want idempotent publishing, I'd way we need to calculate the hash of the crate and compare that instead of the version (so we *actually* know it is the same byte-by-byte crate and *not* a forgotten version bump). Just don't want to hijack the op's issue if that's now what he asked for...

**Comment by weihanglo** [comment]
We cannot do `--skip-already-published` without doing the thing you suggested, especially the forgotten version bump part.

> calculate the hash of the crate and compare that instead of the version (so we actually know it is the same byte-by-byte crate and not a forgotten version bump).

This might be really doable without any new flag added. I can't see a reason people using workspace publish don't want the `--idempotent` or `-skip-already-published` flag.

**Comment by landonxjames** [comment]
>I can't see a reason people using workspace publish don't want the --idempotent or -skip-already-published flag.

+1 to that, I think it should be the default. The AWS SDK has had [logic handling this](https://github.com/smithy-lang/smithy-rs/blob/052538d6a60f933841bbf8a71b298ffcc1fec4cb/tools/ci-build/publisher/src/subcommand/publish.rs#L63-L64) for several years as part of our publishing tool.

My one worry would be the increased traffic on crates.io from the proposed equivalence checks. We have already caused throttling on crates.io with our publishing (see https://github.com/rust-lang/crates.io/issues/11588). That issue has been resolved, but I would hate for the traffic increase to cause any more issues.

**Comment by ehuss** [comment]
> This might be really doable without any new flag added. I can't see a reason people using workspace publish don't want the `--idempotent` or `-skip-already-published` flag.

Would there still be the option of adding a flag that doesn't check the hash? For example, checking the hash won't work for cargo's own workflow which has [this check](https://github.com/rust-lang/cargo/blob/02c807618d4fa1f21ff386197a79105cef62c359/publish.py#L39-L50) which would be nice to remove.

**Comment by weihanglo** [comment]
Sorry @ehuss, I didn't get the connection between checking tarball checksum and publish.py. Can't publish.py be replaced entire with `cargo publish --workspace` plus <https://github.com/rust-lang/cargo/pull/16012>? It might need some `--exclude` for sure, but `cargo publish` already checks registry index existence.

**Comment by ehuss** [comment]
I believe there will be files that change that are mostly inconsequential. For example, `Cargo.lock`, or normalization changes in `Cargo.toml`. Our existing checks won't handle those. We especially don't want publishing to fail, since it won't be caught until release day, and we may not notice at all since it is buried in a log file nobody looks at.

(We also don't catch workspace inheritance changes, though arguably those may be things we want to bump the version for.)

I also think that you may not want to *force* a version bump in all circumstances. Let's say I change a README in my project, but I don't want to publish a new version just for that. My workflow is to usually leave the version number unchanged, and then decide when I make a release which packages to actually publish by bumping their versions.

**Comment by weihanglo** [comment]
Understood, thanks!

The idempotency I understood here requires you to publish packages from the exact same source without any modification. For example, If you don't check in Cargo you are on your own journey, or you need to ensure registry index stay the same with `--offline`, because the input changed and broke idempotency.

Changing README but don't want to publish it is a new workflow than the idempotent workspace publish to me. Cargo doesn't know what the requirement is to skip a package safely. Don't check hash is a escape hatch but like you said there are more configurable points we might want to consider.

BTW, I guess Cargo's workflow won't be affected if the proposed solution get accepted, because publish.py always publish from the exact identical source for all packages, and [bump-check xtask](https://github.com/rust-lang/cargo/blob/02c807618d4fa1f21ff386197a79105cef62c359/crates/xtask-bump-check/src/xtask.rs?plain=1#L361-L397) requires every crate has a version bump even with just single line changed in README.

**Comment by ehuss** [comment]
> BTW, I guess Cargo's workflow won't be affected if the proposed solution get accepted, because publish.py always publish from the exact identical source for all packages, and [bump-check xtask](https://github.com/rust-lang/cargo/blob/02c807618d4fa1f21ff386197a79105cef62c359/crates/xtask-bump-check/src/xtask.rs?plain=1#L361-L397) requires every crate has a version bump even with just single line changed in README.

If we leave `publish.py` the same, I don't think it will be affected by #16012. But it also means we won't be able to use `publish --workspace` (and similar to my other projects, it won't help me and I continue to need to use a manual approach).

`bump-check` doesn't really do the same thing, either. It ignores changes to `Cargo.lock` (which is usually good), but it misses changes to workspace inheritance (which is usually bad). It also forces a bump when changing things like README or a test, even though we don't really need (or want) a bump for those.

Even with `bump-check`, we would not be able to switch to `publish --workspace`.

**Comment by weihanglo** [comment]
Now I see it. Thanks for the explanation!

So basically there are at least two workflows when people want to skip already published packages

1. Idempotent workspace publish from the exactly identical source
	* Useful when you failed to publish in the middle and want to retry.
	* Usually everything bumps its version and is published in lockstep.
	* You don't want to skip publish just because you forgot to bump version for a package.
	* rust-lang/cargo#16012 should take care of this
2. Several workspace publish from different commits/changes in the workspace
	* Workspace members are loosely coupled or completely independent.
	* One package may get version bumped but the others don't but still what to skip those that version didn't changed.
	* The rust-lang/cargo#16012 doesn't help this, because Cargo.lock or workspace inheritance may change the content of packages having no version bump and fail the publish
	* It is hard to find a safe way to determine whether a package is safe to skip beyond checking version in registry index.

The original request seems to mostly focus on the first one, so I did rust-lang/cargo#16012. However, we might want to consider future support for the second workflow, or at least ensure that when it comes into play, we won’t block our way to implement it.

**Comment by epage** [comment]
How many different workflows should we support?  How do we decide which ones we should support?

Also, if we're just now hearing about people wanting checksum comparisons, then maybe this all needs more baking in third-party release management tools.

**Comment by weihanglo** [comment]
> Also, if we're just now hearing about people wanting checksum comparisons, then maybe this all needs more baking in third-party release management tools.

Nah. I thought of the same solution when thinking again the issue of workspace publish being not atomic, and then realized @MOZGIII has already proposed the same thing. We want to have a reliable way to resume for a broken workspace publish, especially for network errors. Checksum comparison is just a guardrail to ensure we're resuming from the same state.

Even for small workspaces once your workspace publish Job failed, you cannot publish through CI again and can only manually publish. As a result you will lose [trusted publishing](https://crates.io/docs/trusted-publishing). So you either do whatever [AWS SDK is doing](https://github.com/rust-lang/cargo/issues/13397#issuecomment-3331940633), or don't use `cargo publish --workspace` at all and instead use third-party release tool.

**Comment by epage** [comment]
> you cannot publish through CI again and can only manually publish. As a result you will lose [trusted publishing](https://crates.io/docs/trusted-publishing).

When we discussed trusted publishing going in, the idea was that it wasn't more trusted than direct publishing, but more trusted than other forms of CI publishing.  I worry if we're framing direct publishing as a negative alternative to trusted publishing.

Also, I feel like if a release process fails, you will have to do manual cleanup either way; a flag would just help simplify this.

Also, I'm not concerned if people aren't using `--workspace` itself; there are many workflows where it wouldn't be applicable.  What is valueable is multiple-package publishing within the workspace.

**Comment by weihanglo** [comment]
> Also, I'm not concerned if people aren't using `--workspace` itself; there are many workflows where it wouldn't be applicable. What is valueable is multiple-package publishing within the workspace.

Putting `--workspace` as first step is just for making it less controversial (but apparently it didn't 😓). multiple-package publishing actually will also benefits from the idempotent behavior. In the wild large workspace use `cargo publish` a lot and have reimplemented the same `already_published` check. For examples

* Eclipse Zenoh [reimplemented the `already_published` check](https://github.com/eclipse-zenoh/ci/blob/8a210ad113bbd3fae462e9bf937499d05794cad4/src/publish-crates-cargo.ts#L175-L181). So do [oxc](https://github.com/oxc-project/cargo-release-oxc/blob/59d4d04f580e9675cc4352c1fb3f5bc5fa0bb2e2/src/publish.rs#L69-L80) and [cargo-mono](https://github.com/kdy1/cargo-mono/blob/f86305ad6e54024c372427e1f0b1f4a49e5ef85b/src/publish.rs#L81-L91) (monorepo tool swc uses).
* Apache OpenDAL [run `cargo publish` in CI](https://github.com/apache/opendal/blob/fb0cf35dc5cec6650502f5923338727824513ee9/.github/workflows/release_rust.yml#L43-L54). Failed in the middle will require manual publish or temporarily change to the workflow yaml.
* polars has [20+ workspace members](https://github.com/pola-rs/polars/blob/485618ad33eb056028ede4dff809eb993436b529/crates/Makefile#L111-L135) to publish. If they opt-in workspace/multi-package publish and failed in the middle, it is a bit annoying for them to resume, as the need to figure out what hasn't and reconstruct the CLI. So does [Apache arrow-rs](https://github.com/apache/arrow-rs/blob/b444ea7127ebc8136564bc9af036353d0c90991b/dev/release/README.md?plain=1#L242-L264), [windows-rs](https://github.com/microsoft/windows-rs/blob/aafae1f2d445b954f7032dcaf3bd702f85cf5899/docs/publish.cmd).

Yeah if a project is that big, should just use some release management tool for sure, but those tool still need to deal with already_published and retry logic (like AWS Rust SDK). The idempotent feature would help make this part more correct and less re-work on release tool side. Upon this point, maybe `cargo-publish` as a plumbing level command should provide a flag instead of auto opt-in idempotent behavior.

>  I worry if we're framing direct publishing as a negative alternative to trusted publishing.

It is not "negative", but trusted publishing does provides a better and easier provenance tracking than manual publish. Also, while being a bug in cargo, publishing on different platform may get different results especially on Windows.

**Comment by weihanglo** [comment]
Another aspect to consider is what can be done in cargo but hard or impossible in external release tools. If we think tarball checksum verification is useful enough, then it is going to be impossible for external tool to implement that reliably due to the proposed behavior change in <https://github.com/rust-lang/cargo/pull/15915>.

**Comment by MOZGIII** [comment]
Some other ref here: I use https://github.com/pksunkara/cargo-workspaces#publish in the `cargo workspaces publish --publish-as-is` form my monorepos, but it has issues with calculating dependencies relations (and a related `--no-remove-dev-deps` flag for that); I was actually hoping for the `cargo publish --workspace` to cover my needs - but for my case - bumping & publishing only a single package out of many - it doesn't work.

**Comment by ijackson** [comment]
> In the wild large workspace use `cargo publish` a lot and have reimplemented the same `already_published` check. For examples
> 
>   * Eclipse Zenoh [reimplemented the `already_published` check](https://github.com/eclipse-zenoh/ci/blob/8a210ad113bbd3fae462e9bf937499d05794cad4/src/publish-crates-cargo.ts#L175-L181). So do [oxc](https://github.com/oxc-project/cargo-release-oxc/blob/59d4d04f580e9675cc4352c1fb3f5bc5fa0bb2e2/src/publish.rs#L69-L80) and [cargo-mono](https://github.com/kdy1/cargo-mono/blob/f86305ad6e54024c372427e1f0b1f4a49e5ef85b/src/publish.rs#L81-L91) (monorepo tool swc uses).

 * In Arti we have a [publication script](https://gitlab.torproject.org/tpo/core/arti/-/blob/main/maint/cargo-publish?ref_type=heads) one of whose principal functions is to skip already-published crates.  In practice, this is often needed (although maybe not now that #13530 may be fixed?)

This is likely the tip of the iceberg. It surely makes sense to centralise this functionality in cargo.

**Comment by epage** [comment]
I'm concerned that #16012 is implicitly skipping.  imo this should be an explicit selection by the user.

> > Also, I'm not concerned if people aren't using `--workspace` itself; there are many workflows where it wouldn't be applicable. What is valueable is multiple-package publishing within the workspace.
> 
> Putting `--workspace` as first step is just for making it less controversial (but apparently it didn't 😓).

I was not speaking to any specific implementation like #16012 (I didn't know it was only for `--workspace`) but the point being made made it sound like multi-package publishing is pointless without being able to support this workflow with `--workspace` which I disagree with.

> Yeah if a project is that big, should just use some release management tool for sure, but those tool still need to deal with already_published and retry logic (like AWS Rust SDK). The idempotent feature would help make this part more correct and less re-work on release tool side. Upon this point, maybe cargo-publish as a plumbing level command should provide a flag instead of auto opt-in idempotent behavior.

I've not looked at their details but my experience with other release tools involves the release management tool needing to know what is being released.   For them to leverage `--idempotent`, it would need to work in dry-run mode and we'd need a programmatic interface.  That would also be a very expensive operation to perform, making all of that work ... not worth it because it wouldn't be used.

**Comment by nvizzari** [comment]
Apologies if I missed some points while reading through the issue, would the following make sense?

- By default: preserve the existing behavior (fail if a crate has already been published)
- Add `--skip-already-published` for `cargo publish`: this would (only) check for the _existence_ of a crate, and not try to publish if it already exists
- Add `--skip-already-published-verify-checksum` for `cargo publish` (it's quite verbose but someone can suggest a better name): this would check for the _existence_ of a crate, and verify that the checksums match (this would be functionally equivalent to `--idempotent`)

This means:

- the "skip" behaviors are opt-in, which avoids the concerns around forgetting to bump a version
- the checksum comparison can be implemented separately if/when we think this is valuable

----

As an additional data point, my use case is the following: I recently switched from `cargo workspaces publish` to `cargo publish --workspace`, which fails when a version has already been published. This usually happens either because someone re-pushed the same commit (the version is derived from the commit), or because `publish` failed mid-way, and someone retried their CI job. `git commit --allow-empty -m "Republish"` is enough to fix the CI pipeline, but it would be nice not to have to think about it!

**Comment by epage** [comment]
It can be easy to say "let's add a flag for every mode we need" but then they become part of cargo's permanent behavior that we must maintain.  Its especially problematic when they are in the CLI because that is the front and center of our UX and the more we have present there, the harder it is to discover features, the less value people get out of all of the features we provide.

As a counter point to how to design this, `cargo release` has an `--unpublished` flag that is like `--workspace` but filtered to what isn't published.

> As an additional data point, my use case is the following: I recently switched from cargo workspaces publish to cargo publish --workspace, which fails when a version has already been published. 

How are you handling version bumps in manifests and other files, changlogs, tagging, etc?

**Comment by nvizzari** [comment]
Agreed, hence why I suggested to start with `--skip-already-published` as it is (arguably?) the less controversial, but I appreciate even it can feel odd, and you might not want to maintain it/expose it to users.

As far as I am concerned, an `--unpublished` would work as well.

> How are you handling version bumps in manifests and other files, changlogs, tagging, etc?

To publish a new version, we have a script that:
- creates a commit that updates the version in the manifest, or all manifests of the workspace (and possibly other files)
- creates a [release](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases) and a corresponding tag with the new version; _the release_ contains the changelog, we don't have any changelog files in the repository which would require bookkeeping

We also publish ephemeral versions for each commit on a branch (the version is something like `1.0.0-branch.<commit sha>`). Here we automatically update the version in all manifest before publishing, but we don't actually commit back the version change to the branch (it would be annoying).

All crates in the workspaces are always updated together at the same version, which simplifies a lot of things, but we'd like to change this at some point, and it will introduce new issues with versions already published.

**Comment by epage** [comment]
So you had to determine what needs to be published, why can't that be used for determining which packages to select?  What is an `--idempotent` flag buying you?

**Comment by MOZGIII** [comment]
I'd think the hash-check should be active by default, with an opt-in to force sending the packages to the create.io even when the hashes match. Also, on the crates.io side, it would make sense to distinguish between an conflicting package for an already-existing version and the same package for already-existing version; and for the latter case to return a different response (possible a non-error one) indicating to the tooling that is is safe to proceed with publishing further packages.

That kind of first-party support from crates.io would hint all other implementors that this is the right way to handle scenarios where a bunch of packages are to be published or verified; of course - the clients would be encouraged to save bandwidth of the registry and check the hashes even before sending the package over...

**Comment by nvizzari** [comment]
> So you had to determine what needs to be published

Not really, we always try to publish _all_ crates of the workspace (with `cargo publish --workspace` and previously with `cargo workspaces publish`) in CI. `--idempotent` would let us retry publish CI jobs that fail because of intermittent network errors, or permissions errors.

**Comment by Sytten** [comment]
This is really needed, its a PITA right now to publish in CI if it fails for any reason mid-way.
If you don't want to call it idempotent, call it `--skip-existing` or something like it.

**Comment by eglif-roche** [comment]
Our use case is a slight variation of the ones mentioned above.
We have a workspace, containing multiple packages with non-cyclic dependencies. These packages are individually versioned. We use [knope](https://knope.tech/) for bumping the versions.
Our pipeline is responsible for publishing all packages, whose versions have increased, to a JFrog Artifactory (with a Bearer Token).
The pipeline does not know which packages have changed or in which topological order the packages have to be published.

Ideally we could just execute `cargo publish --workspace --skip-existing` to publish all packages which were not already uploaded. We would not need a checksum check, as our pipeline ensures semantic versioning of affected packages.

Other crates like `cargo-release` or `cargo-workspaces` already support skipping existing packages, but they struggle with alternative registries which require authentication ([#670](https://github.com/crate-ci/cargo-release/issues/670), [#175](https://github.com/pksunkara/cargo-workspaces/issues/175)). So, currently we are a bit stuck.

**Comment by jschwe** [comment]
What is the status, now that https://github.com/rust-lang/cargo/pull/16012 was closed? In its current state `cargo publish --workspace` feels somewhat useless, since you can't rerun it if something failed in the middle (network error, forgot to enable trusted publishing for a crate, ...).

**Comment by Mirko-von-Leipzig** [comment]
This would be great if it was supported intrinsically. We currently work around this by first querying the registry for possible crate's existence.

```
cargo info --quiet --registry crates-io "${crate}@${version}"
```

and then build this into a list of `--exclude <crate>` to append to the publish command.

This works, but feels a bit involved for what should imo be the default behavior. Also requires rebuilding the crate list, and a bit of duplicate publish work.
